This repository contains code and data for the manuscript "Evaluating the use of ABBA-BABA statistics to locate introgressed loci" by Simon H. Martin (<shm45@cam.ac.uk>), John W. Davey (<jd626@cam.ac.uk>) and Chris D. Jiggins, available on bioRxiv.

The results in the paper were generated as follows. If details are missing or incorrect, please tell us.

# Simulations

The model scripts should be general enough to be used for your own models. If they are not, please let us know.

Models can be generated using the `shared_ancestry_simulator.R` script. A single combined model can be generated like this:

```
./shared_ancestry_simulator.R -w 10000 -T 60 -c Alternate_t123-0.4_t23-0.2.yml:0.1,Background_t123-0.6_t21-0.4.yml:0.9
```

This will generate 10000 windows, 10% of which will be generated using the model described in the file `Alternate_t123-0.4_t23-0.2.yml` and 90% of which will be generated using `Background_t123-0.6_t21-0.4.yml`, using 60 threads. See the model files folders for the YAML files generated for this paper. The CSV files for the models will be made available in a Data Dryad repository on publication and can be made available on request.

These YAML files were generated by the `run_model_combinations.pl` script, using the following commands for the 5kb, 10kb and 20kb simulations:

```
./run_model_combinations.pl -T 60 -n 10000 -w 5000 -p 1000000 -m File_S6_Model_parameter_list.csv -s path/to/simulator
./run_model_combinations.pl -T 60 -n 5000 -w 10000 -p 1000000 -m File_S6_Model_parameter_list.csv -s path/to/simulator
./run_model_combinations.pl -T 60 -n 2500 -w 20000 -p 1000000 -m File_S6_Model_parameter_list.csv -s path/to/simulator
```

Option `-p` defines the effective population size and option `-m` specifies a file containing the list of models to simulate (available in the repository). Option `-s` is the directory path to the simulator script, `shared_ancestry_simulator.R`. If `-s` is not provided, `run_model_combinations.pl` assumes `shared_ancestry_simulator.R` is in the same folder as itself.

Summary statistics for the models found in the `partition.summary` and `dxy.summary` files were generated as follows:

```
./generate_summary_statistics.R -m model_files_win10000_size5000 -T 60
./generate_summary_statistics.R -m model_files_win5000_size10000 -T 60
./generate_summary_statistics.R -m model_files_win2500_size20000 -T 60
```

# Real Heliconius data

Real data for the Heliconius genome, split into 5 kb windows, can be found in the files `Heliconius_autosome_windows.csv` and `Heliconius_Zchromosome_windows.csv`.

Calculation of summary statistics was done with the python script `egglib_sliding_windows.py`, which makes use of the EGGLIB library. Inputs were "calls" format files, which will be made available in the Dryad repository, or upon request. Window size is specified with the `-w` flag, sliding increment with the `-i` flag and minimum number of sites with the `-m` flag. The latter is a hard cutoff, and windows with fewer sites are discarded. There is also a soft cutoff between 0 and 1, specified with `--minimumExploitableData`. The script will output a column called sitesOverMinExD. At a value of 0.5, this would report the number of sites in the window that had genotype calls for at least 50% of the individuals. Also, the function that reports the number of segregating sites, S, only considers sites with this proportion of individuals covered or greater. For the Z chromosome, ploidy was specified using the `--ploidy` flag, because there were two females in the dataset of [Martin et al. 2013](http://dx.doi.org/10.1101/gr.159426.113). See the command below.

The commands run to calculate values across the colour pattern intervals were:

```
python egglib_sliding_windows.py -i HmB.calls -o Heliconius_HmB_windows.csv -w 5000 -m 1000 -s 1000 -a ABBABABA,dxy,S  -p "P1[Melpomene_aglaope.09-246,Melpomene_aglaope.09-267,Melpomene_aglaope.09-268,Melpomene_aglaope.09-357];P2[Melpomene_amaryllis.09-332,Melpomene_amaryllis.09-333,Melpomene_amaryllis.09-79,Melpomene_amaryllis.09-75];P3[Timareta_Peru.09-312,Timareta_Peru.8624,Timareta_Peru.8628,Timareta_Peru.8631];O[Numata_elegans.1277,Numata_silvana.09-364,Numata_tarapotensis.Hntar05-1358,Pardalinus_sergestus.09-326,Pardalinus_ssp.09-387,Elevatus.09-343,Ethilla.09-63,Hecale.09-345,Ismenius.ismC-mj10]" --report 10 --minimumExploitableData 0.5
```

```
python egglib_sliding_windows.py -i HmYb.calls -o Heliconius_HmYb_windows.csv -w 5000 -m 1000 -s 1000 -a ABBABABA,dxy,S  -p "P1[Melpomene_aglaope.09-246,Melpomene_aglaope.09-267,Melpomene_aglaope.09-268,Melpomene_aglaope.09-357];P2[Melpomene_amaryllis.09-332,Melpomene_amaryllis.09-333,Melpomene_amaryllis.09-79,Melpomene_amaryllis.09-75];P3[Timareta_Peru.09-312,Timareta_Peru.8624,Timareta_Peru.8628,Timareta_Peru.8631];O[Numata_elegans.1277,Numata_silvana.09-364,Numata_tarapotensis.Hntar05-1358,Pardalinus_sergestus.09-326,Pardalinus_ssp.09-387,Elevatus.09-343,Ethilla.09-63,Hecale.09-345,Ismenius.ismC-mj10]" --report 10 --minimumExploitableData 0.5
```

Commands to calculate values for the whole genome were, for autosomes:

```
python egglib_sliding_windows.py -i Heliconius_autosomes  -o Heliconius_autosome_windows.csv -w 5000 -m 1000 -s 5000 -p "P1[ag108,ag572,ag112,ag569];P2[am216,am160,am48,am293];P3[tiP86,tiP313,tiP84,tiP57];O[hec273,eth67,ser202,par371]" -a S,popS,px,dxy,ABBABABA --minimumExploitableData 0.5
```

And for the Z chromosome:
```
python egglib_sliding_windows.py -i Heliconius_Zchromosome  -o Heliconius_Zchromosome_windows.csv -w 1000 -m 1000 -s 5000 -p "P1[ag108,ag572,ag112,ag569];P2[am216,am160,am48,am293];P3[tiP86,tiP313,tiP84,tiP57];O[hec273,eth67,ser202,par371]" -a S,popS,px,dxy,ABBABABA --minimumExploitableData 0.5 --ploidy 2,2,2,2,2,1,2,1,2,2,2,2,2,2,2,2
```

Summary statistics were then generated for the real data:
```
./generate_summary_statistics.R -r Heliconius_autosome_windows.csv
./generate_summary_statistics.R -r Heliconius_Zchromosome_windows.csv
```

# Figures

The figures were generated from the real data and `partition.summary` and 'dxy.summary` files, using the following scripts:

```
./Figure_2.R -i model_files_win10000_size5000/Alternate_t123-1.2_t23-0.2.Background_t123-2_t21-1.2.10000.csv
./Figure_3.R
./Figure_4.R
./Figure_5.R
./Figure_6_S5.R
./Figure_7.R -i model_files_win10000_size5000.dxy.summary.tsv
./Figure_S1_S2.R -i model_files_win10000_size5000.partition.summary.tsv
./Figure_S3_S4.R
```
